Do not try to set invalid duration for deriving due date

oliverguenther opened the pull request:
https://community.openproject.org/work_packages/63598

oliverguenther committed:
Do not try to set invalid duration for deriving due date

https://community.openproject.org/work_packages/63598

oliverguenther commented on spec/services/work_packages/set_attributes_service_spec.rb:
```diff
@@ -1058,6 +1058,46 @@
+            expect(work_package.start_date)
+              .to eq(Time.zone.yesterday)
+            expect(work_package.due_date)
+              .to be_nil
```
This currently assigns `nil` to the due_date. We could arguably return the previous due_date instead? It does not really matter, as the contract rejects the change after all.

cbliard replied:
It's true that it does not really matter as the contract rejects the change, but on the modal, the end date disappears which feels wrong. Besides, on a manually scheduled work package, the end date would remain if the duration is 0, as demonstrated by the test you added and these additional tests:

https://github.com/opf/openproject/blob/083f939fe35ce88233405cbdd8a2349f26b05289/spec/services/work_packages/set_attributes_service_spec.rb#L1040-L1082

and this screenshot:

![image](https://github.com/user-attachments/assets/ab7120cb-f480-4b18-957a-1aca40eff201)

So I think it would be more consistent to keep the due date here too.
I'll fix it.

cbliard committed:
Add proper test for successor with 0 duration

cbliard reviewed (commented):
The integration test you added was failing. I fixed it.

Apart from that, it works. I have a minor remark for which I'll issue a fix.

cbliard committed:
Keep due date when duration is 0 or negative

And order `reuse_current_due_date?` and `duration_assignable?` in the
order they are used.
